exec-service-wasmer/src/wasmer_opcode_control.rs

```diff
@@ -0,0 +1,200 @@
+use std::{
+    mem,
+    sync::{Arc, Mutex},
+};
+
+use loupe::{MemoryUsage, MemoryUsageTracker};
+use wasmer::{
+    wasmparser::Operator, ExportIndex, FunctionMiddleware, GlobalInit, GlobalType, Instance,
+    LocalFunctionIndex, MiddlewareError, MiddlewareReaderState, ModuleMiddleware, Mutability, Type,
+};
+use wasmer_types::{GlobalIndex, ModuleInfo};
+
+use crate::wasmer_breakpoints::{Breakpoints, BREAKPOINT_VALUE_MEMORY_LIMIT};
+
+const OPCODE_CONTROL_MEMORY_GROW_COUNT: &str = "opcode_control_memory_grow_count";
+const OPCODE_CONTROL_OPERAND_BACKUP: &str = "opcode_control_operand_backup";
+
+#[derive(Clone, Debug, MemoryUsage)]
+struct OpcodeControlGlobalIndexes {
+    memory_grow_count_global_index: GlobalIndex,
+    operand_backup_global_index: GlobalIndex,
+}
+
+#[derive(Debug)]
+pub(crate) struct OpcodeControl {
+    max_memory_grow: usize,
+    max_memory_grow_delta: usize,
+    breakpoints_middleware: Arc<Breakpoints>,
+    global_indexes: Mutex<Option<OpcodeControlGlobalIndexes>>,
+}
+
+impl OpcodeControl {
+    pub(crate) fn new(
+        max_memory_grow: usize,
+        max_memory_grow_delta: usize,
+        breakpoints_middleware: Arc<Breakpoints>,
+    ) -> Self {
+        Self {
+            max_memory_grow,
+            max_memory_grow_delta,
+            breakpoints_middleware,
+            global_indexes: Mutex::new(None),
+        }
+    }
+}
+
+unsafe impl Send for OpcodeControl {}
+unsafe impl Sync for OpcodeControl {}
+
+impl MemoryUsage for OpcodeControl {
+    fn size_of_val(&self, tracker: &mut dyn MemoryUsageTracker) -> usize {
+        mem::size_of_val(self) + self.global_indexes.size_of_val(tracker)
+            - mem::size_of_val(&self.global_indexes)
+    }
+}
+
+impl ModuleMiddleware for OpcodeControl {
+    fn generate_function_middleware(
+        &self,
+        _local_function_index: LocalFunctionIndex,
+    ) -> Box<dyn FunctionMiddleware> {
+        Box::new(FunctionOpcodeControl {
+            max_memory_grow: self.max_memory_grow,
+            max_memory_grow_delta: self.max_memory_grow_delta,
+            breakpoints_middleware: self.breakpoints_middleware.clone(),
+            global_indexes: self.global_indexes.lock().unwrap().clone().unwrap(),
+        })
+    }
+
+    fn transform_module_info(&self, module_info: &mut ModuleInfo) {
+        let mut global_indexes = self.global_indexes.lock().unwrap();
+
+        let mut create_global_index = |key: &str, init: i64| {
```

Comment:
Consider creating a method called `create_global_index()`, taking `module_info`, `key` and `init` as parameters. An anonmyous function seems out-of-place here.

Reply:
done